### PR TITLE
feature: add parameter to select specific lines from gist

### PR DESCRIPTION
**Pull request by botoxparty**
Hey!

I was using this plug-in but needed the functionality to select specific lines so thought i'd share it back into the codebase.

It works by adding a `lines` parameter to the query string. e.g.

`gist:botoxparty/04f9c2bd909ba6f4424386d42242f961#serverless.yml?highlights=6,56&lines=1-4,50-60`

RE: Tests, it still passes but I think something has changed with the `isValid` check. What was the purpose of that function?

Do you have any time to write a test for this extended functionality?

Cheers!
Adam

### CODE REVIEW

**Comment by weirdpattern** [comment]
Hi @botoxparty, 

Thanks for sharing this! can you send me an example where you are using this? 
I want to see how it looks, does it just hide the lines?

**Comment by botoxparty** [comment]
Sure, i made a post here with a demo:

https://adamham.dev/posts/gatsby-remark-embed-gist/

It removes the lines from the DOM completely, do you think it should maybe indicate to the user that there are lines removed? I guess that should probably be a separate PR.

image

File: src/index.js [hunk 1]
```diff
@@ -128,7 +143,7 @@ export default async ({ markdownAST }, options = {}) => {
     // highlight the specify lines, if any
     let html = content.div;
     if (query.highlights.length > 0) {
-      const $ = cheerio.load(content.div);
+      const $ = cheerio.load(html);
```
Comment (weirdpattern) [inline src/index.js:146]: I think we can do some optimizations around this area.
Loading html using cheerio is a costly operation, doing it twice is probably just too much.

My suggestion would be:
```js
if (query.highlights.length > 0 || query.lines.length > 0) {
   const $ = cheerio.load(html);
   const file = query.file.replace(/[^a-zA-Z0-9_]+/g, "-").toLowerCase();
   
   // I would do lines first, this actually removes lines from the html
   // no need to highlight lines that are going to be removed anyway =)
   if (query.lines.length > 0) {
     // do removal logic
   }

   if (query.hightlights.length > 0) {
     // do highlighting logic
   }

   html = $.html();
}
```

Comment (weirdpattern) [reply src/index.js:146]: You can move the length counts to variables if you want to - micro-optimization, might not really be needed, but depends on your OCD level - mine is super high 😄)

**Comment by weirdpattern** [comment]
Ah this is great and thanks for the mention and link in your blog post!

I think adding an indicator would be a good idea, but I agree with you, this can be done as a separate PR. 

In the mean time, I submitted a code review, overall great work, just one small optimization.

Let me know what you think 😄

File: src/index.js (hunk 1 above)
Comment (botoxparty) [reply src/index.js]: Awesome, have made those optimisations in the latest commit :)

**Comment by botoxparty** [comment]
So i've made those optimisations in the latest commit. I should probably write tests for the new functionality.

I'll try and follow what you've done with the testing for highlights.

**Comment by weirdpattern** [comment]
Hey @botoxparty just checking on you. Are you planning on adding anything else (tests) to this PR?

Let me know 😄

**Comment by botoxparty** [comment]
hey @weirdpattern 

so i've written out some tests based on the ones you had already written for the highlights. I'm not really sure if I've done it correctly I don't have a lot of experience with testing.

would you mind reviewing it?

**Comment by weirdpattern** [comment]
@botoxparty sorry for the delay.
It all looks good. I'm merging now and publishing tonight

**Pull request merged**